native/common/jp_bridge.cpp

```diff
@@ -0,0 +1,210 @@
+		return;
+	}
+
+	// No guarantees that the exception will make it back so print it first
+	PyObject *err = PyErr_Occurred();
+	if (err != NULL)
+	{
+		PyErr_Print();
+		env->ThrowNew(runtimeException, "Exception in Python");
+	} else
+	{
+		env->ThrowNew(runtimeException, ex.getMessage().c_str());
+	}
+}
+
+PyObject* jpype = nullptr;
+PyObject* jpypep = nullptr;
+
+
+JNIEXPORT void JNICALL Java_org_jpype_bridge_Native_start
+(JNIEnv *env, jobject engine)
+{
+	try
+	{
+		PyStatus status;
+		PyConfig config;
+		PyConfig_InitPythonConfig(&config);
+		status = PyConfig_SetBytesString(&config, &config.program_name, "jpython");
+		if (PyStatus_Exception(status))
+		{
+			PyConfig_Clear(&config);
+			fail(env, "configuration failed");
+		}
+			
+		// Get Python started
+		PyImport_AppendInittab("_jpype", &PyInit__jpype);
+		status = Py_InitializeFromConfig(&config);
+		PyConfig_Clear(&config);
+		if (PyStatus_Exception(status))
+		{
+			fail(env, "Python initialization failed");
+		}
+	
+#if  PY_VERSION_HEX<0x030a0000
+		PyEval_InitThreads();
+#endif
+		s_ThreadState = PyThreadState_Get();
+
+		// Import the Python side to create the hooks
+		jpype = PyImport_ImportModule("jpype");
+		if (jpype == NULL)
+		{
+			fail(env, "jpype module not found");
+			return;
+		}
+		Py_DECREF(jpype);
+
+		// Next install the hooks into the private side.
+		jpypep = PyImport_ImportModule("_jpype");
+		if (jpypep == NULL)
+		{
+			fail(env, "_jpype module not found");
+			return;
+		}
+		PyJPModule_loadResources(jpypep);
+		Py_DECREF(jpypep);
+		
+		// Then attach the private module to the JVM
+		JPContext* context = JPContext_global;
+		context->attachJVM(env);
+		
+		// Initialize the resources in the jpype module
+		PyObject *obj = PyObject_GetAttrString(jpype, "_core");
+		PyObject *obj2 = PyObject_GetAttrString(obj, "initializeResources");
+		PyObject *obj3 = PyTuple_New(0);
+		PyObject_Call(obj2, obj3, NULL);
+		Py_DECREF(obj);
+		Py_DECREF(obj2);
+		Py_DECREF(obj3);
+		
+		// Everything is up and ready
+
+		// Next, we need to release the state so we can return to Java.
+		m_State1 = PyEval_SaveThread();
+	} catch (JPypeException& ex)
+	{
+		convertException(env, ex);
+	}	catch (...)
+	{
+		fail(env, "C++ exception during start");
+	}
+}
+
+
+// This section will allow us to look up symbols in shared libraries from within
+// Java.  We will then call those symbols using the invoker. 
+
+// FIXME global
+static std::list<void*> libraries;
+
+// FIXME this was the full CFFI interface so we could directly call symbols in C.  
+// Likely is not necessary.  But I need it for testing now.
+/*
+ * Class:     python_lang_PyEngine
+ * Method:    getSymbol
+ * Signature: (Ljava/lang/String;)J
+ */
+JNIEXPORT jlong JNICALL Java_org_jpype_bridge_Native_getSymbol
+(JNIEnv *env, jobject, jstring str)
+{
+	try
+	{
+		jboolean copy;
+		const char* name = env->GetStringUTFChars(str, &copy);
+		for (std::list<void*>::iterator iter = libraries.begin();
```

Comment:
I can't tell if this is leftovers that still needs to be refactored but I couldn't find the Java native method declaration.

Also, the access to the `libraries` list is not thread safe. I know msvc does thread safe static globals by default (which is really stupid to create a mutex for every global by default and silently add lock accesses) but I don't know what gcc or clang does. I would like to think they are sane and you would encounter a problem here.

Sorry for going on a minor tangent, I can't help myself sometimes.